SYCL/Matrix/joint_matrix_sub_d.cpp

```diff
@@ -0,0 +1,179 @@
+     auto accB = bufB.get_access<access::mode::read_write>(cgh);
+
+     cgh.parallel_for<class imatrix>(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [=](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]]
+
+         {
+           // The submatrix API has to be accessed by all the workitems in a
+           // subgroup these functions will be called once by the subgroup no
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           ext::oneapi::sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<bfloat16, TM, TK> sub_a(sg);
+           // For B, since current implementation does not support non-packed
+           // layout, users need to specify the updated VNNI sizes along with
+           // the packed_b layout. By default, the layout is row_major and size
+           // is (TK, TN).
+           joint_matrix<bfloat16, TK, TN, matrix_layout::packed_b> sub_b(sg);
+           joint_matrix<float, TM, TN> sub_c(sg);
+           joint_matrix<float, TM, TN> sub_d(sg);
+
+           joint_matrix_load(sg, sub_c,
+                             accC.get_pointer() + (sg_startx * TM) * N +
+                                 sg_starty / SG_SZ * TN,
+                             N, matrix_layout::row_major);
+           joint_matrix_fill(sg, sub_d, 0);
+           joint_matrix_load(sg, sub_a,
+                             accA.get_pointer() + (sg_startx * TM) * K + k * TK,
+                             K, matrix_layout::row_major);
+           // Assuming B data is already in VNNI format.
+           joint_matrix_load(sg, sub_b,
+                             accB.get_pointer() + (k * TK / 2) * (N * 2) +
+                                 sg_starty / SG_SZ * TN * 2,
+                             N * 2, matrix_layout::packed_b);
+           sub_d = joint_matrix_mad(sg, sub_a, sub_b, sub_c);
+           joint_matrix_store(sg, sub_d,
+                              accC.get_pointer() + (sg_startx * TM) * N +
+                                  sg_starty / SG_SZ * TN,
+                              N, matrix_layout::row_major);
+         }); // parallel for
+   }).wait();
+}
+
+static constexpr size_t MATRIX_M = TM * 2;
+static constexpr size_t MATRIX_N = TN * 2;
+static constexpr size_t MATRIX_K = TK;
+bfloat16 A[MATRIX_M][MATRIX_K];
+bfloat16 B[MATRIX_K / 2][MATRIX_N * 2];
+unsigned short Aref[MATRIX_M][MATRIX_K];
+unsigned short Bref[MATRIX_K / 2][MATRIX_N * 2];
+float C[MATRIX_M][MATRIX_N];
+float D[MATRIX_M][MATRIX_N];
+
+float make_fp32(short x) {
+  unsigned int y = x;
+  y = y << 16;
+  float *res = reinterpret_cast<float *>(&y);
+  return *res;
+}
+
+unsigned short make_bf16(float x) {
```

Comment:
I've just remembered that recently it was made possible to call `bfloat16::from_float` from host, so actually I don't think make_bf16 is necessary in these tests anymore.

Reply:
Oh yes and I think you can replace `make_fp32` with `bfloat16::to_float` or the implicit conversion to float.

Reply:
I am still getting: 
terminate called after throwing an instance of 'cl::sycl::exception'
  what():  Bfloat16 conversion is not supported on host device
Aborted (core dumped)

when I try to use from_float

Reply:
Maybe you have to merge the latest sycl branch?

Reply:
Note there seems to still be discussion of whether these are wanted: https://github.com/intel/llvm/pull/6524#discussion_r968390888

Reply:
I am using the nightly internal build from yesterday so it is pretty recent. 
Since the internal branch did not get these changes yet, I would prefer we leave them as it is now.
Once we have a final thing on the host, we can update all the related affected tests in a separate PR.

Reply:
OK, I don't understand why it is missing though. It had it prior to the last commit which was a fix: https://github.com/intel/llvm/commit/66e469ba2c92352733dc84e62995c60ba974eb07